dclab/rtdc_dataset/core.py

```diff
@@ -135,15 +137,20 @@ def __iter__(self):
             yield col
 
     def __len__(self):
+        if self._length is None:
+            self._length = self._get_length()
+        return self._length
+
+    def _get_length(self):
         # Try to get length from metadata.
         length = self.config["experiment"].get("event count")
-        if length:
+        if length is None:
             return length
         # Try to get the length from the feature sizes
-        keys = list(self._events.keys())
+        keys = list(self._events.keys()) + self.features_basin
```

Comment:
Could you please add a separate case for self.features_basin here?
I fear that accessing this property might take very long and might not be required.

i.e.

```python
keys = list(self._events.keys())
for kk in keys:
    ...

keysb = self.features_basin
if keysb:
    return len(self[keysb][0])


raise ValueError...
```

The iteration over the `self._events` is just a precaution due to some internal features not implementing len properly. But for basin features this should not happen.

Reply:
When a dataset is created without any own features, but only referencing features of its basin, then there are no key-value pairs in self._events, so this would return a length of `0`, despite the basin having multiple entries.
So only if `self._events.keys()` is empty I need to look at the `self.features_basin`-list.

So similar to what you suggest I could first check
```python
keys = list(self._events.keys())
if not keys:
    keys = self.features_basin
for kk in keys:
    . . .
``` 
Would that make sense?

Reply:
As a one-liner I could also do:
```python
keys = list(self._events.keys()) or self.features_basin
```
Here the part after `or` would only be executed if `self._events.keys()` is empty. I think it looks a bit nicer and shorter.
What do you prefer?

Reply:
Yes, the `or` looks nicer.